File: LinkListCycle.java
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : no


/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode detectCycle(ListNode head) {
      ListNode slow = head;
      ListNode fast = head;
      boolean flag = false;
      
      //move slow pointer by one step and fast pointer by 2 step
      while(fast != null && fast.next != null) {
        slow = slow.next;
        fast = fast.next.next;
        if(slow == fast) { //if both the pointer meets then cycle found
          flag = true;
          break;
        }
      }
      if(!flag) return null;
      //find the head of the cycle
      fast = head;
      while(slow != fast) {
        slow = slow.next;
        fast = fast.next;
      }
      return slow;
      
    }
}

File: LinkListCycle.py
'''
time complexity: O(n)
space complexity: O(1)
'''
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def detectCycle(self, head: Optional[ListNode]) -> Optional[ListNode]:
        ptr1 = head
        ptr2 = head
        cycle = False
        while( ptr2 and ptr2.next!=None and ptr2.next.next!=None):
            ptr1 = ptr1.next
            ptr2 = ptr2.next.next
            if(ptr1==ptr2):
                cycle=True
                break
        if(cycle==False):
            return None
        ptr1 = head
        while(ptr1!=ptr2):
            ptr1 = ptr1.next
            ptr2 = ptr2.next

        return ptr1
